### ISSUE TITLE
INTERNAL COMPILER ERROR - caused by an @export flag plus ES6 syntax

### ISSUE THREAD

**Issue by simonsarris** [google/closure-compiler#2468]
> Please report this problem.

First, my JavaScript is a product of the [tsickle](https://github.com/angular/tsickle) project. Specifically I'm trying to get a *very simple* library project working with tsickle and closure. It is three files.

In the below case, I have a small `ClassB` which has a property I want to export so that it the method name is not minified when packaged as a library. So I added:

```
    /**
     * @export
     * @return {?}
     */
    toStringPublic() {
```

It is the existence of this flag that is causing the exception to be thrown.

This error seems to be related to ES6. When `tsickle` is run with a target of `es6`, this exception gets thrown when using the compiler. When I target `es5`, this exception is not thrown, and it compiles normally, and exposes the method as I'd expect.

Below is the command line used, plus all JavaScript in the project.

```
PS C:\mini> java -jar .\closure-compiler.jar --js_output_file=lib.js --compilation_level ADVANCED_OPTIMIZATIONS --formatting=pretty_print --export_local_proper
ty_definitions=true --generate_exports --language_out=ECMASCRIPT5 --dependency_mode=LOOSE  --entry_point="tsout.index" --js "./tsout/**.js"
java.lang.RuntimeException: java.lang.RuntimeException: INTERNAL COMPILER ERROR.
Please report this problem.

Unexpected variable goog
  Node(NAME goog): tsout/classB.js:29:4
    toStringPublic() {
  Parent(GETPROP): tsout/classB.js:29:4
    toStringPublic() {

        at com.google.javascript.jscomp.CompilerExecutor.runInCompilerThread(CompilerExecutor.java:126)
        at com.google.javascript.jscomp.Compiler.runInCompilerThread(Compiler.java:753)
        at com.google.javascript.jscomp.Compiler.compile(Compiler.java:723)
        at com.google.javascript.jscomp.Compiler.compile(Compiler.java:693)
        at com.google.javascript.jscomp.AbstractCommandLineRunner.doRun(AbstractCommandLineRunner.java:1080)
        at com.google.javascript.jscomp.AbstractCommandLineRunner.run(AbstractCommandLineRunner.java:492)
        at com.google.javascript.jscomp.CommandLineRunner.main(CommandLineRunner.java:1898)
Caused by: java.lang.RuntimeException: INTERNAL COMPILER ERROR.
Please report this problem.
```

```JavaScript
// classA.js:
goog.module('tsout.classA'); exports = {}; var module = {id: 'tsout/classA.js'};
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * A newly constructed {\@link ClassA} has no string to show; if it did,
\@class
 */
class ClassA {
    constructor() {
    }
    /**
     * \@ignore
     * @return {?}
     */
    toStringInternal() {
        return "this is a class B instance";
    }
}
exports.ClassA = ClassA;
```

```JavaScript
// classB.js:
goog.module('tsout.classB'); exports = {}; var module = {id: 'tsout/classB.js'};
Object.defineProperty(exports, "__esModule", { value: true });
var classA_1 = goog.require('tsout.classA');
/**
 * A newly constructed {\@link ClassB} has no string to show; if it did,
\@class
 */
class ClassB {
    constructor() {
        this.myClassA = new classA_1.ClassA();
    }
    /**
     * Gets or sets the goob property.
    \@name ClassB#goob
    \@function.
    \@return {string}
     * @return {?}
     */
    get goob() { return this.myClassA; }
    /**
     * @param {?} value
     * @return {?}
     */
    set goob(value) { this.myClassA = value; }
    /**
     * @export
     * @return {?}
     */
    toStringPublic() {
        return "this is a ClassB instance and here's " + this.goob.toStringInternal();
    }
    /**
     * \@ignore
     * @return {?}
     */
    shouldBeRenamed() {
        return "this should not exist";
    }
}
exports.ClassB = ClassB;
function ClassB_tsickle_Closure_declarations() {
    /**
     * \@ignore
     * @type {?}
     */
    ClassB.prototype.myClassA;
    /**
     * \@ignore
     * @type {?}
     */
    ClassB.prototype.myPublicClassA;
}
```

```JavaScript
// index.js:
goog.module('tsout.index'); exports = {}; var module = {id: 'tsout/index.js'};
Object.defineProperty(exports, "__esModule", { value: true });
var classA_1 = goog.require('tsout.classA');
var classB_1 = goog.require('tsout.classB');
//export let myLib: any = {}
let /** @type {?} */ myLib = {};
// very semantic version
myLib['version'] = "lalala";
// runtime API
myLib['ClassA'] = classA_1.ClassA;
myLib['ClassB'] = classB_1.ClassB;
((window))['myLib'] = myLib;

```

**Comment by MatrixFrog** [comment]
If you're using `@export` you need to include the Closure Library in your project, because the compiler rewrites the `@export`-annotated things to calls to `goog.exportSymbol`. The error message should be much clearer on this, I admit.

**Comment by simonsarris** [comment]
Why does it work without error when writing ES5 JavaScript, but fails with ES6?

**Comment by MatrixFrog** [comment]
Good question! Could be due to some quirk of how tsickle works perhaps. cc @alexeagle

**Comment by alexeagle** [comment]
We generally run tsickle only on ES6 outputs, as this is the more optimizable language level. When targetting ES5 I suspect tsickle isn't transforming the code at all.

**Comment by brad4d** [comment]
@MatrixFrog is this a correct summary of this issue?

1. JSCompiler is throwing an exception instead of generating an error when `@export` is used without closure-library being included in the compilation. We should fix this.
2. tsickle is failing to include closure-library code when it is needed. This should be filed as an issue on the angular/tsickle project.

**Comment by simonsarris** [comment]
Hold on. Just because `tsickle` generated the `js` doesn't mean its tsickle's fault. This seems like totally valid JS regardless. Consider the following code:

```JavaScript
class ClassB {
    constructor() {
        this.myClassA = new classA_1.ClassA();
    }
    /**
     * Gets or sets the goob property.
     * @export
     *  The flag above works just fine
     * \@name ClassB#goob
     * \@function.
     * @return {?}
     */
    get goob() { return this.myClassA; }
    /**
     * @export
     *  The flag above works just fine
     * @param {?} value
     * @return {?}
     */
    set goob(value) { this.myClassA = value; }
    /**
     * @export
     * The flag above breaks!
     * @return {?}
     */
    toStringPublic() {
        return "this is a ClassB instance and here's " + this.goob.toStringInternal();
    }
    /**
     * \@ignore
     * This will be completely removed by ADVANCED MODE. (good)
     * @return {?}
     */
    shouldBeRenamed() {
        return "this should not exist";
    }
}
```

When I run:

```
java -jar .\closure-compiler.jar --js_output_file=lib.js --compilation_level ADVANCED_OPTIMIZATIONS --formatting=pretty_print --export_local_proper
ty_definitions=true --generate_exports --language_out=ECMASCRIPT5 --dependency_mode=LOOSE  --entry_point="tsout.index" --js "./tsout/**.js"
```

There are three `@export` statements, one on a getter, one setter, and one on a method. The getter and setter are not generating the error. Only the method is. If I remove the `@export` from the method but keep it on the getter/setter, it compiles, and the output looks as I'd expect (The getter and setter retain their "goob" name) instead of getting minified.

By the way this problem is born out of the deprecation of `@expose`, so I am trying to use `@export` instead. I am writing a class that will be part of a browswer `.js` library so it needs advanced mode to keep its methods and not rename them. Is `@export` the wrong thing to use here? The goog library was never a requirement before...

**Comment by ChadKillingsworth** [comment]
@brad4d We should inject the export methods as a polyfill. We have not had that option in the past, but we do now. That's the direction I [went with Polymer](https://github.com/google/closure-compiler/pull/2338/files#diff-82e4b93d43725e5104536c249edffccc).

@simonsarris You need to include the base.js file from closure library for now (and only that file).

**Comment by simonsarris** [comment]
Since adding `base.js` is an acceptable workaround, I'm gonna close this.

Feel free to re-open if you think it should remain as a reminder to make the error message more clear, @MatrixFrog.
